Add support for blackboard types

bjoernschroeder opened the pull request:
Add type support for blackboard APIs. The types supported are bool, int, unsigned int, double, string & any (for any other type). This means that 
- pml files can use these types
- pml files can specify default values for them 
- APIs will throw meaningful exceptions when types mismatch in set & get API
- blackboard key mapping will also verify the types & throw if there are mismatches

Link to wrike task: https://www.wrike.com/open.htm?id=940589144

bjoernschroeder committed:
add support for blackboard types, use std::variant instead of std::any in blackboard

bjoernschroeder committed:
fix int64_t access

bjoernschroeder committed:
add converter to handle get different for variant and any

bjoernschroeder committed:
remove convertConst struct

bjoernschroeder committed:
reenable initDefaultValues

Dale-Koenig commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -23,25 +93,143 @@ class BlackboardImpl
+    YAML::Node node;
+
+private:
+    friend BlackboardUtil;
```
I think we really have a need to separate the blackboard from implementation at this point.  Ideally implementation with separate header file, that is included by `Blackboard.cpp` and not `Blackboard.h` (maybe pimpl?)

veerajsk replied:
most of the functions are templates but yeah, can look into how to do this. But can we defer to a separate PR?

bjoernschroeder committed:
add support for promoting types from source to target blackboard

bjoernschroeder committed:
fix monostate issues

bjoernschroeder committed:
add support for accessing internal type of std::any values in blackboard

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -23,25 +184,110 @@ class BlackboardImpl
     }
     template <typename T>
-    T& get(const std::string& key)
+    const T& get(const std::string& key)
```
why? This should return `T&`
And, there is no need to implement this method, instead can reuse the const method i.e.
`return const_cast<T&>(static_cast<const BlackboardImpl*>(this)->get<T>(key));`

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -23,25 +184,110 @@ class BlackboardImpl
+            std::string pmlType = getBlackboardValueType(key);
+            if (pmlType == "std::any") {
+                // insert as std::any with type T
+                vals.at(key) = std::any{value};
```
This will cause `get<T>` to fail if `T` is a known type (i.e. if `Find<T, BBValueType>::result` is true). So I guess easiest way is to check in `get<T>(key)` if pml type of key is std::any first & if so use `any_cast` there

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -23,25 +184,110 @@ class BlackboardImpl
+    {
+        // Note: srcKey & targetKey has to be set
+        // mapping succeeds as long as targetType is constructible from srcType (so conversions are supported)
+        std::visit(
```
can you add a TODO here to eliminate compiler warnings? I haven't looked at how to do it yet

veerajsk replied:
tried various ways to change the code to eliminate compiler warnings without success, will leave it as a TODO & suppress warnings for now

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -23,25 +184,110 @@ class BlackboardImpl
+    friend BlackboardUtil;
+    YAML::Node getBlackboardValueNode(std::string key) const
+    {
+        for (const auto& entry : node) {
```
This is rather inefficient because it is called every time a key is set on the blackboard. Instead you could store the pml types of each key when the blackboard is created in an unordered_map

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -111,13 +362,17 @@ class LockedBlackboardRW
 
-    void set(const std::string& key, const std::any& value) { _impl->set(key, value); }
+    template <typename T>
+    void set(const std::string& key, const T& value)
```
there is scope for improvement here. We should ideally have 3 ways to set values depending on how `T` is passed i.e. const lvalue, rvalue & emplace i.e.
`set(key, const T&`
`set(key, T&&`
`emplace(key, Args&&...)`

can do in another PR though because it will require forwarding `T` or `Args` through all the intermediate functions using `std::forward`

veerajsk replied:
will be done in subsequent PR

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -23,25 +184,110 @@ class BlackboardImpl
+                // if type is std::any, wrap into std::any
+                vals.emplace(key, std::any{value});
+            } else {
+                vals.emplace(key, value);
```
should not use `emplace` to set the value, because `emplace` fails if the key already exists on the blackboard

veerajsk reviewed (commented):
need to review tests

bjoernschroeder committed:
fix some review comments

veerajsk commented on alica_engine/include/engine/blackboard/BlackboardBlueprint.h:
```diff
@@ -17,10 +21,12 @@ class BlackboardBlueprint
 private:
-    std::unordered_map<std::string, std::any> vals;
+    std::unordered_map<std::string, BlackboardValueType> vals;
+    YAML::Node node;
```
The blueprint can store the object representation of the pml file (for blackboard section) i.e. an unordered_map of <key, type_name, std::optional<default_value>>. This way the YAML::Node need not be accessed/parsed by all internals of the engine

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,15 +1,180 @@
+template <>
+struct convert<std::monostate>
+{
+    static Node encode(const std::monostate& rhs)
```
can get rid of these methods & instead conditionally enable the compilation of the Parser i.e. using if constexpr

bjoernschroeder committed:
add get with non const return value, use if constexpr for type dependent implementation of parser and get, remove yaml implementation for std::any and std::monostate

bjoernschroeder committed:
fix timing issue

bjoernschroeder committed:
clean up code, add comments, fix some errors

veerajsk committed:
Fix tests

veerajsk committed:
Merge branch 'devel' into bas_bb_types_rebased

veerajsk committed:
WIP

veerajsk committed:
Cleanup

veerajsk committed:
Tests working

veerajsk committed:
Merge branch 'devel' into bas_bb_types_rebased

veerajsk committed:
Fix format

veerajsk committed:
Exceptions thrown from get

veerajsk committed:
Throw more exceptions

veerajsk commented:
@bjoernschroeder please review as well. I cannot add you as you are the creator of the PR

Dale-Koenig commented on alica_dummy_proxy/src/communication/AlicaDummyCommunication.cpp:
```diff
@@ -216,12 +216,12 @@ AlicaDummyCommunication::AlicaDummyCommunication(const alica::AlicaCommunication
         , _isRunning(false)
 {
-    s_modContainer.registerModule(this);
+    // s_modContainer.registerModule(this);
```
What's going on here?

veerajsk replied:
@bjoernschroeder I have the same question actually

Dale-Koenig replied:
It does seem related to https://github.com/rapyuta-robotics/alica/pull/356/files

bjoernschroeder replied:
Yes, during the implementation, the bug we had with the logger appeared more often, so I added the fix we discussed with abhi. It should be equal to your PR dale

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,47 +1,325 @@
     const T& get(const std::string& key) const
     {
-        return std::any_cast<const T&>(vals.at(key));
+        // T must be an exact match to the T used while setting this key (i.e. no type conversions taken into account)
```
this comment is slightly inaccurate for keys that are not in the yaml file & whose type does not match exactly with any type in the variant. More precisely, T should be an exact match to the type stored in the variant.

veerajsk commented on supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp:
```diff
@@ -72,5 +75,47 @@ TEST_F(TestBlackBoard, testNotInheritBlackboardFlag)
     EXPECT_FALSE(alica::test::Util::getBasicPlan(ae, 1692837668719979457, 0)->getInheritBlackboard());
 }
 
+TEST_F(TestBlackBoard, testWithoutPlan)
```
need to add more tests

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,47 +1,325 @@
-        return std::any_cast<T&>(vals.at(key));
+        return const_cast<T&>(static_cast<const BlackboardImpl*>(this)->get<T>(key));
+    }
+    BBValueType& get(const std::string& key) { return _vals.at(key); }
```
can remove these?

bjoernschroeder replied:
These are not used anymore, so I dont see any reason to keep them.

bjoernschroeder commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,47 +1,325 @@
+    template <class T>
+    static const char* getTypeName()
+    {
+        auto typeIndex = BBValueType{T{}}.index() - 1;
```
we can call `getTypeName(typeIndex)` here to avoid code duplication. Because getTypeName(std::size_t index) already accounts for monostate, you dont have to substract 1 from typeIndex then.

bjoernschroeder commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,47 +1,325 @@
+            // Note: index cannot be std::monostate
+            // throws if variant construction fails
+            assert(index > 0);
+            return makeHelper(index, std::make_index_sequence<BB_VALUE_TYPE_NAMES_SIZE + 2>(), std::forward<Args>(args)...);
```
Can you add a comment why + 2 is used here? I think it was because we have to iterate over the 2 additional types in std::variant, monostate and any, right?

veerajsk replied:
yes

ntfshard commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,47 +1,325 @@
+    BlackboardImpl(const BlackboardBlueprint* blueprint)
+    {
+        for (auto it = blueprint->begin(); it != blueprint->end(); ++it) {
+            const auto& [key, keyInfo] = *it;
```
Feels like this 2 lines can be merged

ntfshard commented on alica_engine/src/engine/blackboard/BlackboardUtil.cpp:
```diff
@@ -14,24 +14,26 @@ void BlackboardUtil::setInput(const Blackboard* parent_bb, Blackboard* child_bb,
 void BlackboardUtil::setOutput(Blackboard* parent_bb, const Blackboard* child_bb, const KeyMapping* keyMapping)
 {
-    auto lockedParentBb = LockedBlackboardRW(*parent_bb);
+    const auto lockedParentBb = LockedBlackboardRW(*parent_bb);
```
by-copy?

veerajsk replied:
changed the syntax

ntfshard commented on alica_engine/include/engine/blackboard/BlackboardBlueprint.h:
```diff
@@ -1,26 +1,32 @@
+    void addKey(const std::string& key, const std::string& type, std::optional<std::string> defaultValue)
     {
-        vals.emplace(std::piecewise_construct, std::forward_as_tuple(key), std::forward_as_tuple(std::forward<decltype(args)>(args)...));
+        _keyInfo.emplace(std::piecewise_construct, std::forward_as_tuple(key), std::forward_as_tuple(KeyInfo{.type = type, .defaultValue = defaultValue}));
```
`{.type = type` technically requires c++20

ntfshard commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -104,20 +386,22 @@ class LockedBlackboardRW
     LockedBlackboardRW& operator&=(LockedBlackboardRW&) = delete;
     LockedBlackboardRW(LockedBlackboardRW&) = delete;
 
-    template <class... Args>
-    [[deprecated]] void registerValue(const std::string& key, Args&&... args)
+    template <typename T>
+    const T& get(const std::string& key) const
     {
-        _impl->registerValue(key, std::forward<Args>(args)...);
+        return _impl->get<T>(key);
     }
-
     template <typename T>
     T& get(const std::string& key)
     {
         return _impl->get<T>(key);
     }
```
technically one can be implemented, and other could be made with const_cast. But I guess this is also OK

veerajsk replied:
actually, it is better this way since this is just a wrapper class, so it should just forward the call

bjoernschroeder committed:
Add additional tests for blackboard types PR (#363)

* add bb tests for up and down conversion

* add UnknownType class to tests, add tests for get exceptions

* add tests for setting and initialization, add non constructable helper class, rename test fixture for blackboard

* add set and init test for blackboard

* use EXPECT_THROW for exception testing

* - use template for mapping tests
- use constexpr for type names
- improve exception tests, dont check exception msg
- compare UnknownTypes by using an int value, passed during initialization
- add additional tests for std::any

* add tests setWithConvertibleType and setWithoutSpecifyingType, use default constructor to create default values of base types

* activate tests for different types convertible to int64_t

* format

* remove old code

* Set yaml type for blackboard mapping tests

Co-authored-by: Veeraj S Khokale <veeraj.khokale@rapyuta-robotics.com>

veerajsk committed:
Merge branch 'devel' into bas_bb_types_rebased

veerajsk committed:
Remove commented out code

ntfshard commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,47 +1,330 @@
+            // The value itself is either constructed from args or parsed from args
+            // Note: index cannot be std::monostate
+            // throws if variant construction fails
+            assert(index > 0);
```
index is unsigned type. It can be !=0 and/or can be less some upper bound

veerajsk committed:
Review comments

ntfshard commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,47 +1,330 @@
+    }
+
+    static constexpr const char* BB_VALUE_TYPE_NAMES[] = {"bool", "int64", "double", "std::string"};
+    static constexpr std::size_t BB_VALUE_TYPE_NAMES_SIZE = sizeof(BB_VALUE_TYPE_NAMES) / sizeof(const char*);
```
std::array?

Dale-Koenig commented on supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp:
```diff
@@ -72,5 +118,359 @@ TEST_F(TestBlackBoard, testNotInheritBlackboardFlag)
+{
+    EXPECT_TRUE((checkMapping<bool, bool>(BBType::BOOL, true, BBType::BOOL, true)));
+    EXPECT_TRUE((checkMapping<bool, int64_t>(BBType::BOOL, true, BBType::INT64, 1)));
+    EXPECT_TRUE((checkMapping<bool, double>(BBType::BOOL, true, BBType::DOUBLE, 1.0)));
```
I don't think bool to double (or even int) conversations are something we really want to support

veerajsk replied:
A single line change in 292 will determine the kind of conversions we support. Right now I use `std::is_constructible`. But we could of course use our own custom trait if we want specific conversions to be allowed/disallowed.

Dale-Koenig replied:
As mentioned in comment above, I'm not convinced there is value supporting any conversion other than int->double.  Especially narrowing conversions should not be supported imo

veerajsk replied:
Yeah, I did look for a way to prevent narrowing conversions (that also causes the compiler warnings which had to be suppressed) but it turned out to be quite tricky. 
Also bool -> double is not a narrowing conversion so maybe it needs to be dealt with separately.
Let me look into this a bit & see if I can come up with a suitable trait

veerajsk replied:
I found a solution. Basically I now call the set API here to do the mapping. This implies that mapping & set now support the same kind of type conversions which is desirable. The only type conversions supported are promotions i.e. promotions from smaller integral types to 64 bit ones

Dale-Koenig commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,47 +1,322 @@
+
+    private:
+        template <class T>
+        struct Parser
```
Can be separate PR, but IMO would be better if we can remove the yaml parsing from blackboard and preprocess it.

veerajsk replied:
Could be tricky to do because the types are inferred here. Probably will need sharing all the code which does the type inference between the blackboard & the blueprint.

Dale-Koenig replied:
Could also just disable all conversions at first and enable later if we figure it out

Dale-Koenig commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,47 +1,322 @@
+        template <class... Args, std::size_t... Is>
+        static BBValueType makeHelper(std::size_t index, std::index_sequence<Is...>, Args&&... args)
+        {
+            BBValueType vals[] = {makeBBValueIfIndex<Is>::make(index, std::forward<Args>(args)...)...};
```
std::array

Dale-Koenig commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,47 +1,322 @@
+            template <class... Args>
+            static BBValueType make(std::size_t index, Args&&... args)
+            {
+                // if INDEX == index, make a variant with a value of type that is same as the variant's type at index INDEX, intialized (or parsed) with value
```
```suggestion
                // if INDEX == index, make a variant with a value of type that is same as the variant's type at index INDEX, initialized (or parsed) with value
```

Dale-Koenig commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,47 +1,322 @@
+                        return index == INDEX ? BBValueType{TypeAtIndex{Parser<TypeAtIndex>{}(args)...}} : BBValueType{};
+                    }
+                } else {
+// TODO: figure out a way to disable warnings, caused due to std::visit. Note: really tricky since small changes can cause unexpected behaviour, so best to
```
Ideally we should not allow narrowing conversions:  i.e. do not allow mapping float to int.  In fact I think the only conversion worth maybe supporting is int -> double

veerajsk committed:
Use set API for mapping from src to target blackboard

This implies the same kind of type conversions are supported by mapping
& set API. This additionally avoids the need to suppress warnings in the
make variant helper & prevents conversions between type categories i.e.
bool -> double, int -> double etc. but supports conversions within the
same type category eg. int8_t -> int64_t, float -> double etc

veerajsk committed:
Add support for unsigned integer types

veerajsk committed:
Address minor review comments

veerajsk committed:
Fix format

veerajsk committed:
Merge branch 'devel' into bas_bb_types_rebased

Dale-Koenig commented:
PR description, link wrike task, etc

Dale-Koenig commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,47 +1,331 @@
+public:
+    BlackboardImpl() = default;
+    BlackboardImpl(const BlackboardImpl&) = delete;
+    BlackboardImpl(const BlackboardBlueprint* blueprint)
```
I think parsing the yaml etc should be done in the blueprint, not when constructing the blackboard

veerajsk replied:
ok, I will create another task for this & see how it can be done because the parsing related code is currently tied up in the blackboard class. Wrike task: https://www.wrike.com/open.htm?id=1033598893
